robot_calibration/src/finders/scan_finder.cpp

```diff
@@ -0,0 +1,266 @@
+ * Copyright (C) 2015-2017 Fetch Robotics Inc.
+ * Copyright (C) 2013-2014 Unbounded Robotics Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Author: Michael Ferguson
+
+#include <math.h>
+#include <pluginlib/class_list_macros.h>
+#include <robot_calibration/capture/scan_finder.h>
+#include <sensor_msgs/point_cloud2_iterator.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+PLUGINLIB_EXPORT_CLASS(robot_calibration::ScanFinder, robot_calibration::FeatureFinder)
+
+namespace robot_calibration
+{
+
+const unsigned X = 0;
+const unsigned Y = 1;
+const unsigned Z = 2;
+
+ScanFinder::ScanFinder() :
+  tf_listener_(tf_buffer_), waiting_(false)
+{
+}
+
+bool ScanFinder::init(const std::string& name,
+                       ros::NodeHandle & nh)
+{
+  if (!FeatureFinder::init(name, nh))
+    return false;
+
+  // We subscribe to a LaserScan
+  std::string topic_name;
+  nh.param<std::string>("topic", topic_name, "/scan");
+  subscriber_ = nh.subscribe(topic_name,
+                             1,
+                             &ScanFinder::scanCallback,
+                             this);
+
+  // Name of the sensor model that will be used during optimization
+  nh.param<std::string>("sensor_name", laser_sensor_name_, "laser");
+
+  // Frame to transform point cloud into before applying limits below
+  //   if specified as "none", cloud will be processed in sensor frame
+  nh.param<std::string>("transform_frame", transform_frame_, "base_link");
+
+  // It is assumed that the laser scanner operates in the X,Y plane
+  // Valid points must lie within this box, in the laser frame
+  nh.param<double>("min_x", min_x_, -2.0);
+  nh.param<double>("max_x", max_x_, 2.0);
+  nh.param<double>("min_y", min_y_, -2.0);
+  nh.param<double>("max_y", max_y_, 2.0);
+
+  // Repeat points in the Z plane a number of times at fixed distance
+  nh.param<int>("z_repeats", z_repeats_, 10);
+  nh.param<double>("z_offset", z_offset_, 0.1);
+
+  // Should we include debug image/cloud in observations
+  nh.param<bool>("debug", output_debug_, false);
+
+  // Publish the observation as a PointCloud2
+  publisher_ = nh.advertise<sensor_msgs::PointCloud2>(getName() + "_points", 10);
+
+  return true;
+}
+
+void ScanFinder::scanCallback(const sensor_msgs::LaserScan& scan)
+{
+  if (waiting_)
+  {
+    scan_ = scan;
+    waiting_ = false;
+  }
+}
+
+bool ScanFinder::waitForScan()
+{
+  ros::Duration(1/10.0).sleep();
+
+  waiting_ = true;
+  int count = 250;
```

Comment:
I think that it makes more sense to have a "max_wait_time" instead of an int count.

Reply:
this function is straight up stolen from the other finders. I'm going to leave as-is right now, as I plan to eventually refactor these sort of things into the finder base class